### PR TITLE
[Runtimes] Enhance hyper-params, add parallel, early stop, etc.

### PR DESCRIPTION
**Pull request by yaronha**
* use HyperParamOptions in task with new options (limit runs, errors, ..)
* support early stop condition
* support parallel runs using local/remote Dask

### CODE REVIEW

File: docs/hyper-params.ipynb [hunk 1]
```diff
@@ -0,0 +1,1600 @@
+{
```
Comment (Hedingber) [inline docs/hyper-params.ipynb:1]: <p>What is <code>mul</code>  ?</p><p>Maybe give an example that matches the text above, i.e. <code>accuracy</code>  ?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/mlrun/mlrun/pull/791/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (Hedingber) [inline docs/hyper-params.ipynb:1]: <p>you said combinations twice, why not calling it <code>max_combinations</code>  ?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/mlrun/mlrun/pull/791/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Comment (Hedingber) [inline docs/hyper-params.ipynb:1]: <p>Looks like this is mistakenly here </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/mlrun/mlrun/pull/791/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

Comment (Hedingber) [inline docs/hyper-params.ipynb:1]: <p>I suggest to change "level of parallelism" to "concurrency limit"</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/mlrun/mlrun/pull/791/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

Comment (Hedingber) [inline docs/hyper-params.ipynb:1]: <p>Same, suggest to call the param <code>concurrency</code>  <span class="ql-cursor">﻿</span></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/mlrun/mlrun/pull/791/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='21'/>

File: mlrun/__main__.py [hunk 2]
```diff
@@ -98,9 +98,10 @@ def main():
     help="how to select the best result from a list, e.g. max.accuracy",
 )
 @click.option(
-    "--tuning-strategy",
-    default="",
-    help="hyperparam tuning strategy list | grid | random",
+    "--strategy", default="", help="hyperparam tuning strategy list | grid | random",
```
Comment (Hedingber) [inline mlrun/__main__.py:101]: Why changing the name ?

File: mlrun/__main__.py [hunk 3]
```diff
@@ -98,9 +98,10 @@ def main():
...
+)
+@click.option(
+    "--hyper-options", default="", help="hyperparam options json string",
```
Comment (Hedingber) [inline mlrun/__main__.py:104]:
```suggestion
    "--hyper-params-options", default="", help="hyperparam options json string",
```

File: mlrun/__main__.py [hunk 4]
```diff
@@ -145,7 +146,8 @@ def run(
     hyperparam,
     param_file,
     selector,
-    tuning_strategy,
+    strategy,
+    hyper_options,
```
Comment (Hedingber) [inline mlrun/__main__.py:150]: I see this is repeating everywhere, just like that class is `HyperParamOptions` I don't see a reason why it won't be `hyper_param_options` everywhere

File: mlrun/execution.py [hunk 5]
```diff
@@ -248,6 +253,13 @@ def from_dict(
             self._log_level = spec.get("log_level", self._log_level)
             self._function = spec.get("function", self._function)
             self._parameters = spec.get("parameters", self._parameters)
+            if not self._iteration:
+                self._hyperparams = spec.get("hyperparams", self._hyperparams)
+                self._hyper_options = spec.get("hyper_options", self._hyper_options)
+                if isinstance(self._hyper_options, dict):
+                    self._hyper_options = HyperParamOptions.from_dict(
+                        self._hyper_options
+                    )
```
Comment (Hedingber) [inline mlrun/execution.py:262]: Shouldn't you get from the `spec` ?

File: mlrun/kfpops.py [hunk 6]
```diff
@@ -153,7 +154,7 @@ def mlrun_op(
     mode: str = "",
     handler: str = "",
     more_args: list = None,
-    tuning_strategy=None,
```
Comment (Hedingber) [inline mlrun/kfpops.py:156]: Why removing ? this break BC

File: mlrun/model.py [hunk 7]
```diff
@@ -326,6 +326,44 @@ def iteration(self, iteration):
         self._iteration = iteration
 
 
+class HyperParamOptions(ModelObj):
+    """Hyper Parameter Options
+
+    Parameters:
+        param_file (str):       hyper params input file path/url, instead of inline
+        strategy (str):         hyper param strategy - grid, list or random
+        selector (str):         selection criteria for best result ([min|max.]<column>), e.g. max.accuracy
```
Comment (Hedingber) [inline mlrun/model.py:335]:
```suggestion
        selector (str):         selection criteria for best result ([min|max.]<result>), e.g. max.accuracy
```

File: mlrun/model.py [hunk 8]
```diff
@@ -326,6 +326,44 @@ def iteration(self, iteration):
...
+        dask_cluster_uri=None,
+        max_evals=None,
+        max_errors=None,
+        teardown_dask=None,
+    ):
+        self.param_file = param_file
+        self.strategy = strategy
+        self.selector = selector
+        self.stop_condition = stop_condition
+        self.max_evals = max_evals
```
Comment (Hedingber) [inline mlrun/model.py:360]: Create a setter for `max_evals` and explode when it's set and strategy is not `random`

File: mlrun/model.py [hunk 9]
```diff
@@ -378,6 +420,16 @@ def inputs(self):
...
         self._inputs = self._verify_dict(inputs, "inputs")
 
+    @property
+    def hyper_options(self) -> HyperParamOptions:
+        return self._hyper_options
+
+    @hyper_options.setter
+    def hyper_options(self, hyper_options):
+        self._hyper_options = self._verify_dict(
+            hyper_options, "inputs", HyperParamOptions
```
Comment (Hedingber) [inline mlrun/model.py:430]: Change the name, `inputs` is where you copy-pasted from

File: mlrun/runtimes/base.py [hunk 10]
```diff
@@ -409,8 +428,11 @@ def run(
         last_err = None
         if task_generator:
             # multiple runs (based on hyper params or params file)
-            generator = task_generator.generate(runspec)
-            results = self._run_many(generator, execution, runspec)
+            generator = task_generator
```
Comment (Hedingber) [inline mlrun/runtimes/base.py:431]: remove

File: mlrun/runtimes/base.py [hunk 11]
```diff
@@ -789,7 +829,8 @@ def save(self, tag="", versioned=False, refresh=False):
         hash_key = db.store_function(
             obj, self.metadata.name, self.metadata.project, tag, versioned
         )
-        return hash_key
+        hash_key = hash_key if versioned else None
```
Comment (Hedingber) [inline mlrun/runtimes/base.py:832]:
```suggestion
        hash_key = hash_key if versioned else tag or None
```

File: mlrun/runtimes/generators.py [hunk 12]
```diff
@@ -16,57 +16,86 @@
...
 
 
 hyper_types = ["list", "grid", "random"]
 default_max_evals = 10
+default_max_errors = 3
+
+
+def is_hyper_job(spec):
+    param_file = spec.param_file or spec.hyper_options.param_file
+    return param_file or spec.hyperparams
```
Comment (Hedingber) [inline mlrun/runtimes/generators.py:30]: Make this a method of the spec itself

File: mlrun/runtimes/generators.py [hunk 13]
```diff
@@ -16,57 +16,86 @@
...
+    options = spec.hyper_options
+    strategy = spec.strategy or options.strategy
+    if not is_hyper_job(spec) or strategy == "custom":
         return None
 
-    if tuning_strategy and tuning_strategy not in hyper_types:
-        raise ValueError(f"unsupported hyperparams type ({tuning_strategy})")
+    hyperparams = spec.hyperparams
+    param_file = spec.param_file or options.param_file
+    if strategy and strategy not in hyper_types:
```
Comment (Hedingber) [inline mlrun/runtimes/generators.py:41]: Suggest to change `hyper_types` to an enum class (and rename to something more appropriate)

File: mlrun/runtimes/generators.py [hunk 14]
```diff
@@ -16,57 +16,86 @@
...
+        raise ValueError(f"unsupported hyperparams type ({strategy})")
```
Comment (Hedingber) [inline mlrun/runtimes/generators.py:42]:
```suggestion
        raise ValueError(f"unsupported strategy ({strategy})")
```

File: mlrun/runtimes/local.py [hunk 15]
```diff
@@ -40,7 +45,91 @@
...
+            runobj = RunObject.from_dict(resp)
+            try:
+                log_std(self._db_conn, runobj, sout, serr)
+                resp = self._post_run(resp)
+            except RunError as err:
+                resp = self._post_run(resp, err=str(err))
+                num_errors += 1
+            results.append(resp)
+            if num_errors > generator.max_errors:
+                logger.error("too many errors, stopping iterations!")
```
Comment (Hedingber) [inline mlrun/runtimes/local.py:85]:
```suggestion
                logger.error("max errors reached, stopping iterations!")
```

File: mlrun/runtimes/local.py [hunk 16]
```diff
@@ -40,7 +45,91 @@
...
+            stop = generator.eval_stop_condition(run_results)
+            if stop:
+                logger.info(
+                    f"reached early stop condition ({generator.stop_condition}), stopping iterations!"
+                )
+            return stop
+
+        futures = []
+        for task in tasks:
+            # self.store_run(task)
```
Comment (Hedingber) [inline mlrun/runtimes/local.py:97]: remove

File: tests/test_run.py [hunk 17]
```diff
@@ -181,11 +196,31 @@ def test_hyper_list():
...
+
+
 def test_hyper_random():
     grid_params = {"p2": [2, 1, 3], "p3": [10, 20, 30]}
     run_spec = tag_test(base_spec, "test_hyper_random")
-    run_spec.with_hyper_params(grid_params, selector="r1", strategy="random")
-    run_spec.spec.parameters["MAX_RANDOM_EVALS"] = 5
+    run_spec.with_hyper_params(
+        grid_params, selector="r1", strategy="random", max_evals=5
+    )
```
Comment (Hedingber) [inline tests/test_run.py:223]: Add system test for the hyper param with dask

**Review by Hedingber — CHANGES_REQUESTED** [review]

File: mlrun/__main__.py (hunk 2 above)
Comment (yaronha) [reply mlrun/__main__.py:101]: it appears in both names (strategy and tuning-strategy) in different places, so prefer to align all on one name, its hardly used today so not an issue

File: mlrun/__main__.py (hunk 4 above)
Comment (yaronha) [reply mlrun/__main__.py:150]: too long IMO

File: mlrun/execution.py (hunk 5 above)
Comment (yaronha) [reply mlrun/execution.py:262]: the line above (`self._hyper_options = spec.get("hyper_options", self._hyper_options)`) read it from spec

File: mlrun/kfpops.py (hunk 6 above)
Comment (yaronha) [reply mlrun/kfpops.py:156]: wasn't used by the caller anyway (`base.as_step()`)

File: mlrun/model.py (hunk 7 above)
Comment (yaronha) [reply mlrun/model.py:335]: done

File: mlrun/model.py (hunk 8 above)
Comment (yaronha) [reply mlrun/model.py:360]: what if he sets the max eval first and the strategy later? isnt it better to validate in `generator`?

File: mlrun/model.py (hunk 9 above)
Comment (yaronha) [reply mlrun/model.py:430]: done

File: mlrun/runtimes/base.py (hunk 10 above)
Comment (yaronha) [reply mlrun/runtimes/base.py:431]: done

File: mlrun/runtimes/generators.py (hunk 12 above)
Comment (yaronha) [reply mlrun/runtimes/generators.py:30]: done

File: mlrun/runtimes/generators.py (hunk 14 above)
Comment (yaronha) [reply mlrun/runtimes/generators.py:42]: done

File: mlrun/runtimes/local.py (hunk 15 above)
Comment (yaronha) [reply mlrun/runtimes/local.py:85]: done

File: mlrun/runtimes/local.py (hunk 16 above)
Comment (yaronha) [reply mlrun/runtimes/local.py:97]: done

File: mlrun/runtimes/base.py (hunk 11 above)
Comment (yaronha) [reply mlrun/runtimes/base.py:832]: done

Comment (yaronha) [reply mlrun/runtimes/base.py:832]: done

File: mlrun/__main__.py (hunk 2 above)
Comment (yaronha) [reply mlrun/__main__.py:101]: done

File: mlrun/__main__.py (hunk 3 above)
Comment (yaronha) [reply mlrun/__main__.py:104]: done

File: mlrun/__main__.py (hunk 4 above)
Comment (yaronha) [reply mlrun/__main__.py:150]: done

**Review by Hedingber — APPROVED** [review]
